### ISSUE TITLE
The KDE-implementation computes negative density values for some cells

### ISSUE THREAD

**Issue by ahaselsteiner** [ahaselsteiner/compute-hdc#2]
To reproduce this error run computeHdcExampleWithKDE.m .

![image](https://user-images.githubusercontent.com/11452055/69657139-130e9a00-107a-11ea-9093-d9778062bed1.png)

**Comment by BeyondLee518** [comment]
Hello, I was using computeIformContour, but while running it, an error occurred that the output parameters were insufficient. What happened?Also, can I use this package to calculate the environmental contour of wind and  wave?

**Comment by ahaselsteiner** [comment]
Hi @czy-423 

could you provide a minimum example how you used computeIformContour such that I can reproduce the error and try to solve it?

Best, Andreas

**Comment by BeyondLee518** [comment]
Hello, Pro.ahaselsteiner, I am honored to see your code. I wonder if you considered Rosenblatt Transformation when you used IFORM to get the environment contour.Or you can provide some references to the code for IFORM and Rosenblatt Transformation

**Comment by BeyondLee518** [comment]
Hi Andreas,
I have read your paper 'Deriving environmental contours from highest density regions' and I have some questions to ask you.1. How do you simulate Hs-Tp with Monte Carlo?2. In your paper, Hs-Tp of joint probability distribution are used to obtain the environmental contour. Could you change the joint probability distribution of Uw-Hs to obtain the environmental contour?3. As you said in your paper, the environment variable of HDC is 2-p, can it be expanded into 3-p on your basis?Is it easy to operate?

**Comment by ahaselsteiner** [comment]
Hi BeyondLee,

I use standard Monte Carlo simulation. In case you are looking for code examples or how to perform it with the data structure that is used in compute-hdc, you could look at the direct sampling contour implementation: https://github.com/ahaselsteiner/compute-hdc/blob/master/computeDsContour.m

Yes, you can use the highest density method also for a wind speed - wave height joint distribution. You could also use this "compute-hdc" implementation. 

Yes, you could also  use highest density contours for 3-dimensional joint distributions. For example, have a look at this 3D model: https://github.com/ahaselsteiner/compute-hdc/blob/master/getProbabilisticModel.m#L46
Here is a script that runs all of these example models, including 3-dimensional ones: https://github.com/ahaselsteiner/compute-hdc/blob/master/tests/testAllProbabilisticModels.m

There is also an implementation for highest density contours (and other contours) in the Python software package viroconcom that is probably better maintained by us. Might be worth checking out too: https://github.com/virocon-organization/viroconcom
We recently used it for wind-wave contours: https://github.com/ahaselsteiner/2020-paper-omae-hierarchical-models/blob/master/compute_contours_datasets_def.py

**Comment by BeyondLee518** [comment]
Well, thank you for your information!Do you have a manual for running the HDC?I'm running your program of  MATLAB is always wrong, in addition to the 'computeHdc. m', for example, I want to run the examples of 'computeIformContourExample. m', tip error 'undefined function or variable' ExponentiatedWeibull '.ComputeIformContour (Line 104) 
Dist = ExponentiatedWeibull(pm.coeffs {1}{1}, pm.coeffs {1}{2},...
Error computeIformContourExample (line 11)
[x1Array, x2Array] = computeIformContour(PM, alpha, 360, 1);'

**Comment by ahaselsteiner** [comment]
It seems as you do not have the ExponentiatedWeibull class. Did you download the software by using the recommended command (git clone --recurse-submodules https://github.com/ahaselsteiner/compute-hdc.git
)? Alternatively, you can also manually download the ExponentiatedWeibull class. It is available here: https://github.com/ahaselsteiner/exponentiated-weibull/tree/209893872503ffeccffefe82c65bcd1018436a84

Unfortunatley, we don't provide a manual for it (other than the README.md file).

**Comment by BeyondLee518** [comment]
Thank you very much. The problem was solved just now. It is indeed a problem with ExponentiatedWeibull.What you've just run out of is the environmental contour of the wind and the wave, you don't use the raw data, you use the joint probability density function?Now I only have the Weibull distribution of the wind and the probability distribution of wave height and spectral peak period. Can I get the environmental contour by your method?

**Comment by ahaselsteiner** [comment]
Hi BeyondLee518,

I am not 100% sure if I understand your question correctly, but: compute-hdc does not offer functionality to fit a joint distribution to raw data. You would need to write your own code to do that. Then you could use that joint distribution to compute an envronmental contour using compute-hdc. If you already have a joint distribution (a 3D distribution for wind speed, wave height and peak period?), you can create a compute-hdc data structure that represents that distribution and compute a contour based on it.

Since you already posted some questions about our software package viroconcom, I guess you are aware about it. Anyway, viroconcom offers functionality to fit a a joint distribution to raw data / estimate the parameter values of the distribution and then compute a contour, https://github.com/virocon-organization/viroconcom.

Best, Andreas

**Comment by BeyondLee518** [comment]
Yes, Compute -HDC doesn't have the ability to read the data, but I can compute the parameters, such as wave heights according to the three-parameter Weibull distribution and spectrum peak periods according to the lognormal distribution, as you saw in erik Vanema's example in 2012 in HDC, but I find it a little difficult to fit the parameters
